### ISSUE TITLE
RFC: Remove Code Climate from CI processes

### ISSUE THREAD

**Issue by chrisdholt** [microsoft/fast#4492]
This is a fairly straight forward RFC on removing Code Climate from our CI toolkit. Code Climate wasn't working for a bit and it's resurgence has revealed issues which seem to make it *less helpful* overall.

### Why should we remove Code Climate?
1. **Causing more confusion than help** - currently we have regular confusion as to whether code climate issues are *required* for being fixed, why they are coming up, etc. This is confusing for core contributors as well as first-time and external contributors. While I think the automation here can help, it's not without problems...see next...
2. **It doesn't learn from "won't fix" or "dismissals"** - An example here is that certain things in our repository are cause for duplication. In the past, we have consistently marked these as "won't fix", yet Code Climate continues to throw on these.
3. **It is opinionated and is not scoped** - While similar to certain tools like prettier which are opinionated by design, Code Climate targets the entire repo and isn't able to be scoped. Could we scope these, we could target packages, file types, etc that we care about. As it is now, this leads to the issue above.
4. **Admin Permissions* - Somehow it seems like our ability to set or maintain admin permissions have changed. If code owners cannot consistently dismiss or won't fix these, our issues will grow over time here and it will be perceived incorrectly as debt.

While I understand that Code Climate allows us to "measure" certain aspects of quality, the issues we have with it outweigh the benefits for me.

From a principled standpoint, I believe that any automated tooling which helps is a welcome addition to our toolkit. If it ceases to provide the same value or starts to create more problems and maintenance issues, we should remove it. We are taking this approach with lerna for publishing and I believe we should take it here as well.

**Comment by janechu** [comment]
During the time it was disabled, I can't say it was missed. I agree with all the points @chrisdholt raised, especially #3.

This being a mono repository, there are quite a few duplications in particular which are purposeful and necessary to prevent a rats nest. That's just a single example, there are other examples that come to mind and ultimately it creates more work than it helps solve potential issues.

**Comment by SethDonohue** [comment]
I have noticed the comments it adds to PRs and they don't seem very helpful. To reiterate your point, I could see them being confusing to all users, but especially to first-time contributors.

**Comment by awentzel** [comment]
While these are valid concerns, without code climate we have no way to measure and educate around code quality.  Code Quality is the fundamental best method we currently have to help market and advertise how the work we do compares to other open-source projects and our competitors.  Additionally, it provides great tips based on industry best practices for new engineers to learn how to write high-quality code and understanding the impact.

- Technical Debt
- Lines of Code
- Code Maintainability

I would prefer to first file issues we have and raise these as bugs for Code Climate to take action on.  For example, it's a known issue that it doesn't handle mono-repositories well.  I've brought this up in the past and they were working on a solution.  It's also a known issue that it's difficult to manage roles and seemingly no way possible.

Both issues seem reasonable to ask Code Climate to prioritize as a starting point.

**Comment by Garbee** [comment]
>  without code climate we have no way to measure and educate around code quality

Is this really true? Does alternative tooling not exist that could be leveraged to supply the same relative data as Code Climate that could then be acted upon by custom CI logic?

WebStorm's project analysis for one seems to do at least some of the functions of Code Climate (duplicate code chunks across the codebase for example.) It at the very least could be leveraged in some way for CI if no other alternatives exist.

I think before proceeding with outright removal the landscape should be assessed.

* Is what Code Climate providing actually useful (examples of beneficial output)?
* What are the options to replace that in the community?
* What is the estimated long term cost of ownership to maintain those things in other ways?

**Comment by chrisdholt** [comment]
Thanks @awentzel and @garbee for the thoughtful responses here. I think there's good things to consider and I have some thoughts I'll add below:

> Code Quality is the fundamental best method we currently have to help market and advertise how the work we do compares to other open-source projects and our competitors. Additionally, it provides great tips based on industry best practices for new engineers to learn how to write high-quality code and understanding the impact.

I think code quality is important, but I'd argue that we have some fantastic engineers contributing to the project who are catching much of this in their own. In several PR's recently and without the help of code climate, we've had folks call out improvements, offers solutions to simplify, etc. I'm not sure we need code climate to catch those things.

Secondly, Code Climate's suggestions are somewhat subjective in terms of complexity. The other day I saw a PR which called out a cognitive complexity of "1" as an issue. As far as how cognitive complexity is calculated, one needs to go and read a white paper referenced by Code Climate to understand how it's calculated - only then can you make that content somewhat actionable. It seems quite subjective rather than something which is clear, action, and objective feedback - which is helpful as part of the review process.

As far as providing great tips, I don't know that I see that being the case.I think it is maintainers and reviews that serve the above role more effectively. Currently, code climate spits out notes about either cognitive complexity or duplication and new engineers or contributors have commonly assumed that they've done something wrong, when in fact - there isn't an issue there. This causes more confusion than help IMO.

> I would prefer to first file issues we have and raise these as bugs for Code Climate to take action on. For example, it's a known issue that it doesn't handle mono-repositories well. I've brought this up in the past and they were working on a solution. It's also a known issue that it's difficult to manage roles and seemingly no way possible.

I'm fine with filing issues and feature requests, but I don't think we continue to use it as is while we wait for them to prioritize it on their backlog and get pretty significant known issues fiixed.

Regarding @Garbee's feedback :) -

> I think before proceeding with outright removal the landscape should be assessed.
>
> - Is what Code Climate providing actually useful (examples of beneficial output)?
I won't say an outright "no", but I think this has been extremely rare. At the beginning of FAST, with a smaller crew, this provided key hints on potential issues to look for. As our repo and products have matured, I think our need for Code Climate has actually decreased.
> - What are the options to replace that in the community?
I think the community is an immediate drop in replacement. I'm *ALL FOR* finding a replacement technology. As I mentioned above, when something is beneficial we can and should consider adopting it. When it ceases to be beneficial and helpful we should drop it. We're seeing this from a publishing standpoint right now w/ moving away from Lerna towards Beacball.
> - What is the estimated long term cost of ownership to maintain those things in other ways?
Currently, I believe I spend more time mitigating incorrect information each time this runs on a PR than would take to actually *carefully review* the PR. I think we have a net benefit by simply removing it as of now and performing reviews as normal. As mentioned above, I've been pinged several times by people asking how they can fix something that Code Climate has considered an "issue" which is fundamentally by design in our repo. Since I can't dismiss or suppress entire folders, file types, etc from getting these warnings, the issue continually compounds. Thankfully, people have reached out and asked before going and trying to follow through with these "improvements". Doing so would actually make things much worse.

TLDR; I think we're better off tomorrow without Code Climate in its current state. I'd be happy to file feature requests and revisit it once they make those fixes and updates. I'd also be more than happy to identify other solutions available in the space. As it currently stands now, I think Code Climate does more harm than good to our workflow, the perception of the project as it appears on PRs, and our contribution model.

**Comment by EisenbergEffect** [comment]
@awentzel Perhaps we can disable this for now in order to remove the friction for the team. Then take a task to research alternative solutions. These could even be command line tools that we run over specific parts of the codebase as needed. For example, pulling GitHub data or analyzing dependency trees via webpack plugins.

That said, let's make sure you have the data you want for your upcoming DevOps communication before turning it off. Then, we can message as part of that not only what the data shows but what our plans are for future investigation around alternative tools. Basically, I want to make sure the reporting you are planning this cycle isn't sabotaged by this change. So, let's get that report out and message the change, and then switch over to the new plan.

Thoughts?

**Comment by awentzel** [comment]
I've deactivated the display of these informational notices on GitHub.  I'm putting in the backlog to verify these are removed. I also have had a ticket outstanding for a few weeks since this RFC was written and awaiting a response to get more information on usage including additional items I've been curious about.

**Comment by awentzel** [comment]
For 5 months we've had this resolved per my previous comment so I'm closing. Additionally, code climate hasn't made any progress on adding improved support for mono-repositories.  I don't expect they will and so in the future if this becomes problematic again we should investigate what other options may exist in this space.
